### PR TITLE
[PIP-152] Support subscription level dispatch rate limiter setting.

### PR DESCRIPTION
**Pull request by Jason918**
### Motivation

See PIP-152 #15094

### Modifications

See #15094

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
  - org.apache.pulsar.broker.admin.TopicPoliciesTest#testSubscriptionLevelDispatchRate

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (**yes**)
  - The admin cli options: (**yes**)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 
  
- [x] `doc` 
Doc added in admin client api.

### CODE REVIEW

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/TopicPolicies.java [hunk 1]
```diff
@@ -71,6 +71,10 @@ public class TopicPolicies {
     private Integer maxSubscriptionsPerTopic;
     private DispatchRateImpl replicatorDispatchRate;
     private SchemaCompatibilityStrategy schemaCompatibilityStrategy;
+    /**
+     * Dispatch rate limiter setting for specific subscription.
+     */
+    private Map<String/*subscription*/, DispatchRateImpl> subscriptionLevelDispatchRateMap = new HashMap<>();
```
Comment (codelipenghui) [inline pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/TopicPolicies.java:77]: Do we need to introduce SubscriptionPolicies here? We might need to support other policies like backlog quota, TTL.

Comment (codelipenghui) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/TopicPolicies.java:77]: And need to add `@Builder.Default` annotation? details to see https://github.com/apache/pulsar/pull/12009

Comment (Jason918) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/TopicPolicies.java:77]: > Do we need to introduce SubscriptionPolicies here? We might need to support other policies like backlog quota, TTL.

Good point.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java [hunk 2]
```diff
@@ -157,7 +160,13 @@ public SubscribeRate getSubscribeRate() {
         return this.topicPolicies.getSubscribeRate().get();
     }
 
-    public DispatchRateImpl getSubscriptionDispatchRate() {
+    public DispatchRateImpl getSubscriptionDispatchRate(String subscriptionName) {
+        DispatchRateImpl rate = Optional.ofNullable(subscriptionPolicies.get(subscriptionName))
+                .map(SubscriptionPolicies::getDispatchRate).orElse(null);
+        rate = DispatchRateImpl.normalize(rate);
```
Comment (complone) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:166]: Whether it should be judged here that ```rate``` is not empty

**Review by complone — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -4396,6 +4397,60 @@ protected CompletableFuture<Void> internalRemoveSubscriptionDispatchRate(boolean
...
+                SubscriptionPolicies sp = topicPolicies.getSubscriptionPolicies().get(subName);
+                if (sp == null) {
+                    return CompletableFuture.completedFuture(null);
+                }
+                sp.setDispatchRate(null);
+                if (sp.checkEmpty()) {
+                    // cleanup empty SubscriptionPolicies
+                    topicPolicies.getSubscriptionPolicies().remove(subName, sp);
+                }
+                topicPolicies.setIsGlobal(isGlobal);
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:4449]: Why call the `topicPolicies.setIsGlobal(isGlobal)`?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java [hunk 4]
```diff
@@ -1316,6 +1316,56 @@ public void testGetSetSubscriptionDispatchRate() throws Exception {
...
+
+        DispatchRate subLevelRate = DispatchRate.builder()
+                .dispatchThrottlingRateInMsg(1000)
+                .dispatchThrottlingRateInByte(1024 * 1024)
+                .ratePeriodInSecond(1)
+                .build();
+
+        admin.topicPolicies().setSubscriptionDispatchRate(topic, subscriptionName, subLevelRate);
+
+        Assert.assertEquals(admin.topicPolicies().getSubscriptionDispatchRate(topic, subscriptionName), subLevelRate);
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java:1352]: Use `Awaitility`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 3 above)
Comment (Jason918) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:4449]: @nodece Just follow previous pattern. Maybe not necessary, but no harm.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java [hunk 5]
```diff
@@ -1316,6 +1316,56 @@ public void testGetSetSubscriptionDispatchRate() throws Exception {
...
+                abstractTopic.getSubscriptionDispatchRate(subscriptionName), subLevelRate));
+
+        DispatchRateLimiter dispatchRateLimiter = abstractTopic.getSubscription(subscriptionName)
+                .getDispatcher().getRateLimiter().get();
+        Assert.assertNotNull(dispatchRateLimiter);
+        Assert.assertEquals(dispatchRateLimiter.getDispatchRateOnByte(), subLevelRate.getDispatchThrottlingRateInByte());
+        Assert.assertEquals(dispatchRateLimiter.getDispatchRateOnMsg(), subLevelRate.getDispatchThrottlingRateInMsg());
+
+        admin.topicPolicies().removeSubscriptionDispatchRate(topic, subscriptionName);
+        Assert.assertNull(admin.topicPolicies().getSubscriptionDispatchRate(topic, subscriptionName));
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java:1364]: Use `Awaitility`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 2 above)
Comment (Jason918) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:166]: `DispatchRateImpl` don't have concept of `empty`. `normalize` can handle the case when rate is `null`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 3 above)
Comment (nodece) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:4449]: Got

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java (hunk 5 above)
Comment (Jason918) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java]: fixed

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/TopicPolicies.java (hunk 1 above)
Comment (nodece) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/TopicPolicies.java]: You forget to add the `@Builder.Default` here.

**Comment by Jason918** [comment]
@codelipenghui @nodece @complone PTAL

**Review by nodece — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 2 above)
Comment (AnonHxy) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:166]: This code can be written more convenient like this：
```
              return Optional.ofNullable(subscriptionPolicies.get(subscriptionName))
                .map(SubscriptionPolicies::getDispatchRate)
                .map(DispatchRateImpl::normalize)
                .orElse(this.topicPolicies.getSubscriptionDispatchRate().get());
```

**Comment by AnonHxy** [comment]
LGTM with a minor suggestion

**Review by Technoboy- — APPROVED** [review]
LGTM

**Comment by complone** [comment]
LGTM

**Review by complone — APPROVED** [review]
LGTM

**Review by AnonHxy — APPROVED** [review]

**Comment by AnonHxy** [comment]
LGTM

File: pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/SubscriptionPolicies.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+public class SubscriptionPolicies {
+    private DispatchRateImpl dispatchRate;
+
+    /**
+     * Check if this SubscriptionPolicies is empty. Empty SubscriptionPolicies can be auto removed from TopicPolicies.
+     * @return true if this SubscriptionPolicies is empty.
+     */
+    public boolean checkEmpty() {
```
[the remark this answers is not in the document]
Comment (Jason918) [reply pulsar-common/src/main/java/org/apache/pulsar/common/policies/data/SubscriptionPolicies.java:42]: Some serialization lib would use `isXXX` method as boolean properties. 
I prefer using `checkXXX` to indicate it's an action, not a property.

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**